### PR TITLE
[rfc] Add sql specification.

### PR DESCRIPTION
**Pull request by meijies**
The design document for metable supporting single table query statement. It also doesn't support join and subquery statement.

### CODE REVIEW

File: rfcs/000-sql-support.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
+# Proposal: SQL Support
+
+## Status
+
```
Comment (jovany-wang) [inline rfcs/000-sql-support.md:4]: Add Author section?

File: rfcs/000-sql-support.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+Under Discussion
+
+## Background
+
+The sql is the most used way to interact with table store system. Metable need to support it for user accessing data by
```
Comment (jovany-wang) [inline rfcs/000-sql-support.md:9]:
```suggestion
The sql is the most used way to interact with table storage system. Metable needs to support it for user accessing data by
```

File: rfcs/000-sql-support.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+command line tool, web page and client. Further more，we will also introduce extra two actions, there are `subscribe`
```
Comment (jovany-wang) [inline rfcs/000-sql-support.md:10]:
```suggestion
command line tool, web page and client. Further more, we will also introduce 2 extra actions, `subscribe`
```

**Review by jovany-wang — COMMENTED** [review]
At high level, we don't support `publish` because when a table being changed, a event will publish out automatically.

File: rfcs/000-sql-support.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+**supported condition**
+
++ `>`, `<`, `=`, `<=`, `>=`, `!=`
++ `in`
++ `like`, `not like`
++ `between ... and ...`
++ `is null` and `= null`
+
+**2) DML **
```
Comment (jovany-wang) [inline rfcs/000-sql-support.md:68]:
```suggestion
** 2) DML **
```

File: rfcs/000-sql-support.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+delete from <table_name> where <condition> (and/or <condition>)*
+
+select (*|<column_name>) (, *|<column_name>)* from <table_name> where <condition> (and/or <condition>)*
+
+update <table_name> set <column_name> = <column_value> (,<column_name> = <column_value>)* where <condition> (and/or <condition>)*
+```
+
+## command line tool
+
+the command line tool will let user access data easily at server environment. It supports execute sql statement and
```
Comment (jovany-wang) [inline rfcs/000-sql-support.md:83]:
```suggestion
The command line tool will let user access data easily at server environment. It supports executing sql statement and
```

File: rfcs/000-sql-support.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+allow user to set environment variables to controller how to print data, send data and so on.
```
Comment (jovany-wang) [inline rfcs/000-sql-support.md:84]:
```suggestion
allows user setting environment variables to controller how to print data, send data and so on.
```

File: rfcs/000-sql-support.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+**2) How to set environment variables**
+some example here.
+
+```
+set print_format=csv;
+set select_action=subscribe;
+```
+
+## client library
```
Comment (jovany-wang) [inline rfcs/000-sql-support.md:102]: I believe we should remove `client library` section because we're focusing on the sql part in this rfc.

**Comment by jovany-wang** [comment]
How should we `subscribe` in sql? I didn't see it in this rfc.

**Comment by meijies** [comment]
@jovany-wang 
you can review again.

File: rfcs/000-sql-support.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+## Author
+
+meijies
+
+## Background
+
+The sql is the most used way to interact with table storage system. Metable needs to support it for user accessing data
+by command line tool, web page and client. Furthermore, we will also introduce 2 extra actions, `subscribe`and `publish`
+. `subscribe` action use select statement to represent what data do you want and keep fetching snapshot data and
```
Comment (jovany-wang) [inline rfcs/000-sql-support.md:15]:
```suggestion
. `subscribe` action uses select statement to represent what data do you want and keep fetching snapshot data and
```

**Comment by jovany-wang** [comment]
@meijies Actually, I still don't know what the way is to be used for `subscribe` according to your doc. Could you please give the examples?

**Comment by meijies** [comment]
@jovany-wang 

for command line tool, you can subscribe data from table as flow.

> set select_action=subscribe;
> set print_format=csv;
> select * from <table_name>;
==============snapshot data==============
row[1] (column_value, column_value, column_value)
row[2] (column_value, column_value, column_value)
row[3] (column_value, column_value, column_value)
.......
==============incurement data=============
row[n] (column_value, column_value, column_value)
row[n+1] (column_value, column_value, column_value)

the command line tool keep wait data arriving until user interupting the `subscribe`.

**Comment by jovany-wang** [comment]
So why not use a `subscribe` key word like:
```
subscribe * from table_xxx ...;

```

**Comment by meijies** [comment]
> So why not use a `subscribe` key word like:
> 
> ```
> subscribe * from table_xxx ...;
> ```

@jovany-wang 
the client library how to execute subscribe sql statement ? 
``` 
client.sql(<sub_sql>).query(<handler>)
```
or 
```
client.sql(<sub_sql>).sub(<handler>)
```
they don't look very elegant

**Comment by jovany-wang** [comment]
I'll leave the instances later.

**Comment by jovany-wang** [comment]
So I previously want to introduce a function concept for that. Like it's able to define a func to print or do something for result.
```sql
let func_print = [](str name, int age) {
  csv.write_line(name, age);
}

subscribe (name, age) from table_x with func_print;
```

**Comment by jovany-wang** [comment]
Also then func is able to be applied on a data result, like:
```sql 

let fun_incr = [](int x) {x = x + 1;}
let ds = select * from table_x;
ds.map(age, func_inrc);
```

File: rfcs/000-sql-support.md [hunk 9]
```diff
@@ -0,0 +1,104 @@
+# Proposal: SQL Support
+
+## Status
+
+Under Discussion
```
Comment (jovany-wang) [inline rfcs/000-sql-support.md:5]:
```suggestion
FINALIZED
```

File: rfcs/000-sql-support.md [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+| env variable  | description                                                               |
+|---------------|---------------------------------------------------------------------------|
+| print_format  | the format to print selected data, default is table, also support csv     |
+
+**2) How to set environment variables**
+some example here.
+
+```
+set print_format=csv;
+set select_action=subscribe;
```
Comment (jovany-wang) [inline rfcs/000-sql-support.md:103]:
```suggestion
```

File: rfcs/000-sql-support.md [hunk 11]
```diff
@@ -0,0 +1,104 @@
+# Proposal: SQL Support
```
Comment (jovany-wang) [inline rfcs/000-sql-support.md:1]:
```suggestion
# Proposal: Metable SQL-like Specification
```

**Review by jovany-wang — APPROVED** [review]
Looks good to me. A beautiful beginning.

**Pull request merged**